MoreLinq/ScanBy.cs

```diff
@@ -0,0 +1,134 @@
+            Func<TKey, TState> seedSelector,
+            Func<TSource, TKey, TState, TState> accumulator,
+            Func<TSource, TKey, TState, TResult> resultSelector)
+        {
+            return source.ScanBy(keySelector, seedSelector, accumulator, resultSelector, null);
+        }
+
+        /// <summary>
+        /// Scans the source sequence and performs accumulations of state for each group of keys.
+        /// The resulting sequence's elements are generated by projecting each element of the
+        /// source sequece, its key and the current state for that key.
+        /// </summary>
+        /// <typeparam name="TSource">Type of the elements of the source sequence.</typeparam>
+        /// <typeparam name="TKey">Type of the projected key.</typeparam>
+        /// <typeparam name="TState">Type of state elements.</typeparam>
+        /// <typeparam name="TResult">Type of the projected element.</typeparam>
+        /// <param name="source">Source sequence.</param>
+        /// <param name="keySelector">Function that transforms each item of source sequence into a key.</param>
+        /// <param name="seedSelector">Function that defines the initial state for that key.</param>
+        /// <param name="accumulator">Function that defines the current state for that key.</param>
+        /// <param name="resultSelector">The projection to each element, its key and the current state for that key.</param>
+        /// <param name="comparer">The equality comparer to use to determine whether or not keys are equal.
+        /// If null, the default equality comparer for <typeparamref name="TSource"/> is used.</param>
+        /// <returns>A sequence with all projected elements.</returns>
+
+        public static IEnumerable<TResult> ScanBy<TSource, TKey, TState, TResult>(
+            this IEnumerable<TSource> source,
+            Func<TSource, TKey> keySelector,
+            Func<TKey, TState> seedSelector,
+            Func<TSource, TKey, TState, TState> accumulator,
+            Func<TSource, TKey, TState, TResult> resultSelector,
+            IEqualityComparer<TKey> comparer)
+        {
+            if (source == null) throw new ArgumentNullException(nameof(source));
+            if (keySelector == null) throw new ArgumentNullException(nameof(keySelector));
+            if (seedSelector == null) throw new ArgumentNullException(nameof(seedSelector));
+            if (accumulator == null) throw new ArgumentNullException(nameof(accumulator));
+            if (resultSelector == null) throw new ArgumentNullException(nameof(resultSelector));
+
+            return _(); IEnumerable<TResult> _()
+            {
+                comparer = comparer ?? EqualityComparer<TKey>.Default;
+
+                var stateByKey = new Dictionary<TKey, TState>(comparer);
+                var havePrevKey = false;
+                var prevKey = default(TKey);
+                var valueOfNullKey = (Value: default(TState), HasValue: false);
```

Comment:
Better to use tuples or not but not a mix of both; that is, `valueOfNullKey` is a tuple but `havePrevKey` and `prevKey` are maintained in separate variables. So either break out the `valueOfNullKey` tuple like this:

```c#
var haveNullKeyState = false;
var nullKeyState = default(TState);
```

or use tuples defined like this:

```c#
var prevKey = (HasValue: false, Value: default(TKey));
var nullKeyState = (HasValue: false, Value: default(TState));
```